### PR TITLE
Separate lc split

### PR DESCRIPTION
**Pull request by cbartz**
Add option for separate training data split for label calibration

### CODE REVIEW

File: qualle/interface/cli.py [hunk 1]
```diff
@@ -179,13 +180,25 @@ def cli_entrypoint():
...
     train_parser.set_defaults(func=handle_train)
     train_parser.add_argument(
-        "train_data_path",
-        type=str,
-        help="Path to train data. "
+        "predict_train_data_path",
+        type=Path,
+        help="Path to train data for the Recall Predictor. "
         "Accepted inputs are either a tsv file or "
         "a folder in Annif format.",
```
Comment (annakasprzik) [inline qualle/interface/cli.py:187]: Maybe you already answered that for stwfsapy, just wondering again: Do people know what THE Annif format ist? Otherwise, can we include a link here?

File: qualle/interface/cli.py [hunk 2]
```diff
@@ -179,13 +180,25 @@ def cli_entrypoint():
...
     )
     train_parser.add_argument("output", type=str, help="Path to output model file")
+    train_parser.add_argument(
+        "--label-calibration-train-data",
+        "-l",
+        type=Path,
+        dest="label_calibration_train_data_path",
+        help="Path to train data for the Label Calibrator. If omitted, the Label "
+        "Calibrator trained on the same split as the Recall Predictor. "
```
Comment (annakasprzik) [inline qualle/interface/cli.py:196]: "the Label Calibrator is"?

File: qualle/interface/cli.py [hunk 3]
```diff
@@ -179,13 +180,25 @@ def cli_entrypoint():
...
+        "In this case, the required input for the Recall Predictor is estimated "
+        "by cross-validation (sklearn.model_selection.cross_val_predict)."
+        "Accepted inputs are either a tsv file or "
+        "a folder in Annif format.",
```
Comment (annakasprzik) [inline qualle/interface/cli.py:200]: see above

File: README.md [hunk 4]
```diff
@@ -103,13 +103,19 @@ where
 (this is basically the output of the MLC method)
 - ``true_labels`` is a comma-separated list of true labels (ground truth)
 
-Note that you can omit the ``true_labels`` section if you only want to predict the 
+Note that you can omit the ``true_labels`` section if you only want to predict the
```
Comment (annakasprzik) [inline README.md:106]: suggestion: "if you only want to obtain a prediction of the quality of the MLC result."

File: README.md [hunk 5]
```diff
@@ -103,13 +103,19 @@ where
...
 quality of the MLC result.
 
 For example, a row in the data file could look like this:
 
 ``Optimal investment policy of the regulated firm\tConcept0:0.5,Concept1:1\tConcept0,Concept3``
 
+You have the option to provide a separate data split for Label Calibration training
+(see section Train for more information). In this case,
+you must omit the ``predict_labels_with_scores`` section, so that the format looks like
```
Comment (annakasprzik) [inline README.md:115]: "you must omit the ... step"? "... column"?

File: README.md [hunk 6]
```diff
@@ -121,15 +127,27 @@ This is a  folder with three files per document:
 * ``doc.txt`` : document content
 
 As above, you may omit the ``doc.tsv`` if you just want to
-predict the quality of the MLC result.
+predict the quality of the MLC result and you may omit the ``doc.annif`` if you provide
+the split for Label Calibration training.
```
Comment (annakasprzik) [inline README.md:131]: " ... training yourself."?

File: README.md [hunk 7]
```diff
@@ -121,15 +127,27 @@ This is a  folder with three files per document:
...
 
 ### Train
 To train a model, use the ``train`` mode, e.g.:
 
 ``qualle train /path/to/train_data_file /path/to/output/model``
 
-It is also possible to use label calibration (comparison of predicted vs actual labels) using the subthesauri of a thesaurus (such as the [STW](http://zbw.eu/stw/version/latest/about))
-as categories (please read the paper for more explanations). Consult the help (see above) for the required options.
+Qualle uses a feature called Label Calibration, which is a multi-output regression model that provides expectations
```
Comment (annakasprzik) [inline README.md:138]: "an estimate" instead of "expectations"? "expectations" sounds a bit humanized to me but maybe it is the technical term?

File: qualle/interface/cli.py (hunk 1 above)
Comment (cbartz) [reply qualle/interface/cli.py:187]: The tsv and Annif format are explained in the README file. Perhaps we should reference that?

"For more information about the input format, see the README file at https://github.com/zbw/qualle."

File: README.md [hunk 8]
```diff
@@ -121,15 +127,27 @@ This is a  folder with three files per document:
...
+is trained.
+You can provide a separate training data split for Label Calibration:
+
+``qualle train /path/to/train_data_file /path/to/output/model --label-calibration-train-data /path/to/lc_train_data_file``
+
+If a separate training data split is not provided, the Label calibrator will be trained with the same training data split as the
+Recall Predictor, and the required input for training the Recall Predictor is provided by
+[cross-validated estimates](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_predict.html) .
+
+It is also possible to use Label Calibration  using the subthesauri of a thesaurus (such as the [STW](http://zbw.eu/stw/version/latest/about))
```
Comment (annakasprzik) [inline README.md:149]: "It is also possible to apply" to avoid double "use"

File: README.md [hunk 9]
```diff
@@ -103,13 +103,19 @@ where
...
-Note that you can omit the ``true_labels`` section if you only want to predict the 
+Note that you can omit the ``true_labels`` section if you only want to predict the
 quality of the MLC result.
 
 For example, a row in the data file could look like this:
 
 ``Optimal investment policy of the regulated firm\tConcept0:0.5,Concept1:1\tConcept0,Concept3``
 
+You have the option to provide a separate data split for Label Calibration training
+(see section Train for more information). In this case,
```
[the remark this answers is not in the document]
Comment (cbartz) [reply README.md:114]: Why uppercase for "Section" ?

File: README.md [hunk 10]
```diff
@@ -121,15 +127,27 @@ This is a  folder with three files per document:
...
-as categories (please read the paper for more explanations). Consult the help (see above) for the required options.
+Qualle uses a feature called Label Calibration, which is a multi-output regression model that provides expectations
+of the correct number of concepts for a document. This model is trained before the quality estimation regressor (Recall Predictor)
+is trained.
+You can provide a separate training data split for Label Calibration:
+
+``qualle train /path/to/train_data_file /path/to/output/model --label-calibration-train-data /path/to/lc_train_data_file``
+
+If a separate training data split is not provided, the Label calibrator will be trained with the same training data split as the
+Recall Predictor, and the required input for training the Recall Predictor is provided by
```
Comment (annakasprzik) [inline README.md:146]: Just curious: What could be negative effects of using the same training data set for both Label Calibration and Recall Prediction? Is it possible to add one sentence to make the decision easier or at least to make users aware of that? (I understand that on the other hand providing a separate split is only possible if you have enough data.)

**Review by annakasprzik — CHANGES_REQUESTED** [review]
see individual comments

File: README.md (hunk 7 above)
Comment (cbartz) [reply README.md:138]: Actually, I used the wording from the Paper: "Moreover, a multi-output regression
module offers expectations regarding the proper number of concepts for a doc-
ument (Label Calibration)," . But I can also use "an estimate".

File: qualle/interface/cli.py (hunk 1 above)
Comment (annakasprzik) [reply qualle/interface/cli.py:187]: Yes, that would do.

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (annakasprzik) [reply README.md:114]: Because that is how you reference parts of a text, I think? "see Chapter 5, see Subsection 6.1" https://ellii.com/blog/the-capitalization-of-certain-terms-exercise-page-chapter-unit-paragraph-line

File: README.md (hunk 7 above)
Comment (annakasprzik) [reply README.md:138]: no, that's fine then

File: README.md (hunk 10 above)
Comment (cbartz) [reply README.md:146]: So the recall predictor needs input from the label calibrator. If a separate split is not provided, the label calibrator will be trained on the same split. But we can't use this trained calibrator to create the input for the predictor since it's trained on the same data and is therefore overfitted.
So we use cross-validation, which technically means
further label calibrators x_1,...,x_n are used to provide the input. So, in summary, the quality estimation is performed with a different label calibrator than the one used for the training.

In addition, the label calibrators x_1,...,x_n are trained on a part of the train data used by the recall predictor. This means that the recall predictor is not trained on "unseen" data, which can lead to bias.

If you provide a separate split, the overall quality estimate will be cleaner because you are using the same label calibrator in training and in prediction, and the recall predictor has been trained on "unseen data".

It's hard to say for sure whether one approach or the other leads to better results, but I want to give the users of Qualle a choice.

Comment (cbartz) [reply README.md:146]: So, I would suggest to append following text:

"This implies that the quality estimation is not performed with the exact same Label Calibrator used in the training, 
and since the trained calibrators from the cross-validation use parts of the same training data as the Recall Predictor,
the Recall Predictor is not trained on "unseen" data, what could lead to a bias.
Whether one or the other approach has a negative effect must be tested individually."

Comment (annakasprzik) [reply README.md:146]: Okay! I trust that only people who are able to make that call will use qualle anyway.

**Comment by cbartz** [comment]
The option did not lead to better evaluation results for us internally and we have therefore decided to close the PR.
